plutus-core/executables/debugger/Debugger.hs

```diff
@@ -0,0 +1,157 @@
+                BC.center returnValueEditor B.<+> BC.center cekStateEditor
+            , footer
+            ]
+    drawTexts :: [Text] -> B.Widget ResourceName
+    drawTexts =
+        B.vBox
+            . toWidgets
+            -- This is needed for empty lines to be rendered correctly.
+            -- `Brick.Widgets.Core.txt` does the same via `fixEmpty`.
+            . fmap (\t -> if Text.null t then " " else t)
+
+    toWidgets :: [Text] -> [B.Widget ResourceName]
+    toWidgets rows = case st ^. dsUplcHighlight of
+        Nothing -> B.txt <$> rows
+        Just (HighlightCode (B.Location (sRow, sCol)) eLoc) ->
+            let alg :: (Text, Int) -> B.Widget ResourceName
+                alg (row, rowNo)
+                    | rowNo < sRow || rowNo > eRow = B.txt row
+                    -- The current line (either entirely or part of it) needs to be highlighted.
+                    | otherwise =
+                        let s =
+                                if rowNo > sRow
+                                    then 0
+                                    else sCol - 1
+                            e =
+                                if rowNo < eRow
+                                    then Text.length row
+                                    else eCol
+                            -- left -> no highlight
+                            -- middle -> highlight
+                            -- right -> no highlight
+                            ((left, middle), right) = first (Text.splitAt s) (Text.splitAt e row)
+                         in B.hBox $
+                                [B.txt left | not (Text.null left)]
+                                    ++ [ B.withAttr highlightAttr $ B.txt middle
+                                       | not (Text.null middle)
+                                       ]
+                                    ++ [B.txt right | not (Text.null right)]
+                  where
+                    B.Location (eRow, eCol) = fromMaybe (B.Location (sRow, Text.length row)) eLoc
+             in fmap alg (zip rows [1 ..])
+
+-- | Handle events in the debugging stage
+handleDebuggerEvent :: B.BrickEvent ResourceName e -> B.EventM ResourceName DebuggerState ()
+handleDebuggerEvent bev@(B.VtyEvent ev) = do
+    focusRing <- gets (^. dsFocusRing)
+    let handleEditorEvent = case B.focusGetCurrent focusRing of
+            Just EditorUplc ->
+                B.zoom dsUplcEditor $ BE.handleEditorEvent bev
+            Just EditorSource ->
+                B.zoom dsSourceEditor $ BE.handleEditorEvent bev
+            Just EditorReturnValue ->
+                B.zoom dsReturnValueEditor $ BE.handleEditorEvent bev
+            Just EditorCekState ->
+                B.zoom dsCekStateEditor $ BE.handleEditorEvent bev
+            _ -> pure ()
+    case ev of
+        Vty.EvKey (Vty.KChar 's') [] -> modify' $ \st' ->
+            -- Stepping. Currently it highlights one line at a time.
+            let highlightNextLine = \case
```

Comment:
belongs next to `HighlightCode`?

Reply:
This is temporary. Once we have the CEK machine integrated, `highlightNextLine` will be replaced by highlighting the next Term being computed/returned, and I think that logic should live in this module.